aptos-move/aptos-vm/src/data_cache.rs

```diff
@@ -29,20 +111,56 @@ impl<'a, S: StateView> StorageAdapter<'a, S> {
         Self(state_store)
     }
 
-    pub fn get(&self, access_path: &AccessPath) -> PartialVMResult<Option<Vec<u8>>> {
+    pub fn get(&self, access_path: AccessPath) -> PartialVMResult<Option<Vec<u8>>> {
         self.0
-            .get_state_value(&StateKey::AccessPath(access_path.clone()))
+            .get_state_value(&StateKey::AccessPath(access_path))
             .map_err(|_| PartialVMError::new(StatusCode::STORAGE_ERROR))
     }
 }
 
+impl<'a, S: StateView> MoveResolverExt for StorageAdapter<'a, S> {
+    fn get_module_metadata(&self, module_id: ModuleId) -> Option<RuntimeModuleMetadataV1> {
```

Comment:
Not sure whether this an issue, but let me note that the content the resolver sees is not always the same as the VM. After an upgrade, the VM still sees the old, while storage has new. I _think_ we should be good, because we already flushing the cache on every code upgrade.